src/gc/stringobjhashtable.cpp

```diff
@@ -0,0 +1,246 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "stringobjhashtable.h"
+#include "gcenv.h"
+#include "gcenv.inl"
+
+
+bool GCHashTableBase::Init(uint32_t numBuckets, AllocationHeap heap)
+{
+    buckets = new (nothrow) GCHashEntry*[numBuckets]();
+
+    if (buckets == NULL)
+        return false;
+
+    this->numBuckets = numBuckets;
+    this->heap = heap;
+    
+    return true;
+}
+
+StringDupsList** GCHashTableBase::InsertOrGetValue(GCStringData* key, uint32_t hash)
+{
+    assert(numBuckets != 0);
+
+    uint32_t bucket = hash % numBuckets;
+    GCHashEntry* search;
+
+    for (search = buckets[bucket]; search; search = search->Next)
+    {
+        if (search->HashValue == hash && CompareKeys(search, key))
+        {
+            // why one would insert same address twice?
+            assert(search->Key.GetAddress() != key->GetAddress());
+            // key already holds "original" pointer
+            // we are saving space here and allocating dups list only when first duplicate appears
+            // StringDedupTable::Insert need to be aware of such behavior
+            if (!search->Data)
+            {
+                StringDupsList* list = new (nothrow) StringDupsList;
+                if (!list)
+                {
+                    return NULL;
+                }
+                search->Data = list;
+            }  
+            return &(search->Data);
+        }
+    }
+
+    if  (numEntries > numBuckets*2)
+    {
+        if (!GrowHashTable())
+        {
+            return NULL;
+        }
+        bucket = hash % numBuckets;
+    }
+
+    GCHashEntry* newEntry = AllocateEntry(key);
+    if (!newEntry)
+    {
+        return NULL;
+    }
+
+    newEntry->Next = buckets[bucket];
+    newEntry->Data = NULL;
+    newEntry->HashValue = hash;
+
+    buckets[bucket] = newEntry;
+
+    numEntries++;
+
+    return &(newEntry->Data);
+}
+
+uint32_t GCHashTableBase::GetHash(GCStringData* key)
+{
+    uint32_t hash = 5381;
```

Comment:
I'd imagine we'd want to use a cryptographic hash if we want to avoid potential denial-of-service problems here.

Reply:
String itnern pool routines uses this algorithm.
https://github.com/dotnet/coreclr/blob/master/src/vm/eehash.cpp#L289
https://github.com/dotnet/coreclr/blob/master/src/inc/utilcode.h#L3020
And I just simply copied it.

Reply:
@morganbr

Reply:
My intuition feels like this hash function is OK for string interning. Strings that get interned are only attacker-controlled when passed verbatim to `string.Intern` (assuming our threat model doesn't consider string literals in metadata to be attacker controlled). In this case, though, any string on the heap (and therefore passing through this code path) is potentially attacker controlled and arbitrarily large. I'd be worried of attackers repeatedly colliding strings and forcing the slow string comparison path on potentially large numbers of arbitrarily large strings.

Reply:
@swgillespie could you please explain this:
> <85kb since LOH strings aren't eligible for this

I'm agree with your concerns, but keep in mind that this code meant to happen concurrent with mutators, i.e. not affecting STW pause. Also stronger hash function may affect performance more than collision-induced long chains of entries. And finally such malicious strings have to survive probably several gen0 and following gen1 collections.
I'm not against your suggestion, just want all trade-offs to be considered.
And last, we can try different structures like trie-based ones and trade hashing issues for cache thrashing issues 🤣 . Of course this too exotic, just for the record.

Reply:
> collision-induced long chains of entries

You can give up adding strings to the hashtable when you get collision-induced long chain.

Reply:
@Rattenkrieg Correct me if I'm wrong, but my understanding of your approach is that strings only get deduped when they are promoted into gen 2, correct? In that case, LOH strings won't ever get promoted into gen 2 (since they are *allocated* in Gen2/LOH) and won't trigger this logic. Is that correct?

Reply:
@swgillespie that's right! I've completely missed the way how large strings instantly get promoted. Though I don't think there are going to be duplicates among large strings, towards the finish of this PR we may consider knobs to probe large heap too.

Reply:
Creating collisions with this algorithm and a hash table with power-of-two buckets is trivial -- you just need to have the last few characters of the string be the same. We should assume an attacker does have the ability to create a bunch of strings and have them last for a little while. That's not particularly far-fetched when they can submit things like a json document with lots of strings in it to a server.

We've taken two approaches to this that would be acceptable:
1. Always use a randomly seeded strong hash. In the VM, we use https://github.com/dotnet/coreclr/blob/32f0f9721afb584b4a14d69135bea7ddc129f755/src/vm/marvin32.cpp . Marvin is about half the speed of algorithms like HashBytes (though it may make up some of that with a better distribution of results, especially in the power-of-two hash table). I'd recommend testing with this to see if it's good enough to always use (it's how we implement String.GetHashCode).
2. Detect attacks and change strategies. We do that in Dictionary by counting collisions and switching to Marvin if we hit a threshold. You might also be able to detect attacks and just skip putting the colliding strings in the table, but that would have a couple of issues:
a. Strings would still have to be hashed/checked to see that they are more collisions. That might still add some load, especially if they're long.
b. If a service depends on the improved performance of deduplication, an attack that disables it might still act as a denial-of-service.